src/services/StreamPeriodsService.ts

```diff
@@ -0,0 +1,190 @@
+	);
+
+	// Map stream periods into virtualized periods based on conf
+	return streamPeriods.map((streamPeriod) => {
+		const tokenPriceData = tokensWithPriceData.find(
+			(tokenWithPriceData) =>
+				tokenWithPriceData.chainId === streamPeriod.chainId &&
+				tokenWithPriceData.token.toLowerCase() === streamPeriod.token.underlyingAddress.toLowerCase(),
+		);
+
+		return {
+			...streamPeriod,
+			virtualPeriods: virtualizeStreamPeriod(
+				streamPeriod,
+				fromUnixTime(startTimestamp),
+				fromUnixTime(endTimestamp),
+				period,
+				tokenPriceData?.prices || [],
+			),
+		};
+	});
+}
+
+/**
+ * Recursive function to virtualize stream period in different time spans. (Daily, Weekly, Monthly etc.)
+ * Amounts are calculated based on the overlapping period of date filter, stream period and virtualized period.
+ */
+function virtualizeStreamPeriod(
+	streamPeriod: StreamPeriodResult,
+	startDate: Date,
+	endDate: Date,
+	period: UnitOfTime,
+	priceData: TimespanPrice[],
+) {
+	const { flowRate, startedAtTimestamp, stoppedAtTimestamp } = streamPeriod;
+
+	const streamStoppedTimestamp = stoppedAtTimestamp || getUnixTime(Date.now());
+	const endTimestamp = Math.min(getUnixTime(endDate), streamStoppedTimestamp);
+
+	// Date when stream period and virtual period start overlapping
+	const streamPeriodStartTimestamp = Math.max(getUnixTime(startDate), startedAtTimestamp);
+
+	// Virtual period start and end timestamps (start and end timestamps of day, week, month or year)
+	const virtualPeriodEndTimestamp = getEndOfPeriodTimestamp(streamPeriodStartTimestamp, period);
+
+	// Timestamp when stream period, virtual period end and end date filter stop overlapping
+	const streamPeriodEndTimestamp = Math.min(virtualPeriodEndTimestamp, endTimestamp);
+
+	const virtualStreamPeriod: VirtualStreamPeriod = {
+		startTime: streamPeriodStartTimestamp,
+		endTime: streamPeriodEndTimestamp,
+		amount: getAmountInTimespan(streamPeriodStartTimestamp, streamPeriodEndTimestamp, flowRate).toString(),
+		amountFiat: calculateVirtualStreamPeriodPrice(
+			streamPeriodStartTimestamp,
+			streamPeriodEndTimestamp,
+			flowRate,
+			priceData,
+		).toString(),
+	};
+
+	if (endTimestamp <= virtualPeriodEndTimestamp) return [virtualStreamPeriod];
+
+	const nextPeriodStartDate = fromUnixTime(virtualPeriodEndTimestamp + 1);
+	return [
+		virtualStreamPeriod,
+		...virtualizeStreamPeriod(streamPeriod, nextPeriodStartDate, endDate, period, priceData),
+	];
+}
+
+function getAmountInTimespan(startTimestamp: number, endTimestamp: number, flowRate: string): Decimal {
+	return new Decimal(flowRate).mul(new Decimal(endTimestamp - startTimestamp));
+}
+
+function getUniqueNetworkTokenAddresses(streamPeriods: StreamPeriodResult[]): NetworkToken[] {
+	return Object.values(
+		streamPeriods.reduce((tokens, streamPeriod) => {
+			const {
+				chainId,
+				token: { id, underlyingAddress },
+			} = streamPeriod;
+
+			// TODO: If no underlyingAddress then it is native super token?

```

Comment:
Pure super tokens or native asset super tokens. Couple of native asset super tokens have an underlying token marked as well (MATICx comes to mind which will point to WMATIC). Subgraph has `isNativeAssetSuperToken` mapped so if it's not true and there's no underlying then it's a pure super token.